# Description

Machine learning algorithm using Elastic-Net regression.

## Import Libraries

In [1]:
import pandas as pd # library for data manipulation (dataframes)
import numpy as np # library for computations
import matplotlib.pyplot as plt # library for visualization
from sklearn.linear_model import ElasticNet  # library for machine learning methods
from sklearn.linear_model import ElasticNetCV# library for machine learning methods
import xlsxwriter as xlsw # library for exporting excel files

## Definition for Importing Data

In [2]:
def import_data(file_location, sheet):
    """This function returns data read from an xlsx file from a specific table.
    The sheet inside the file must include a '#' as the origin of the table.
    """
    #read data table from file
    values = pd.read_excel(file_location, sheet_name=sheet)
    #set origin of table at '#'
    values.set_index('#', inplace=True)
    return values

## Import Trainning Data

Assign address for trainning data table to variable ('x' values).
This set represents quantitative data for a group of units.
In this example the units are buildings and the data for each
building are NORMALIZED measurements such as areas, connectivity
between spaces, visibility inside spaces, etc.

In [3]:
x_values = import_data('quant_data_example.xlsx', 'trainning_set')

Print data table.

In [4]:
x_values

Type01Area  Type01Integration  Type01Entropy  Type01Control  \
#                                                                          
building 01      0.3268             0.8130         0.6042         0.2189   
building 02      0.5315             1.0000         0.2517         0.3264   
building 03      1.0000             0.8907         0.8582         0.5267   
building 04      0.6187             0.6493         0.6915         0.2182   
building 05      0.3755             0.4120         0.4309         0.2839   
building 06      0.1771             0.7367         1.0000         1.0000   

             Type01Choice  Type01IsoArea  Type01IsoPerim  Type01IsoOclu  \
#                                                                         
building 01        0.7927         1.0000          1.0000         0.6138   
building 02        0.2048         0.1912          0.3008         1.0000   
building 03        0.7560         0.4752          0.5688         0.3545   
building 04        0.2445         0.2112          0.2746         0.0178   
building 05        0.2430         0.1067          0.1492         0.0008   
building 06        1.0000         0.4612          0.6002         0.2352   

             Type02Area  Type02Integration  ...  Type05IsoPerim  \
#                                           ...                   
building 01      0.2250             0.8849  ...          0.1151   
building 02      0.5730             0.6173  ...          0.0088   
building 03      0.4487             0.5113  ...          0.1483   
building 04      0.5990             0.6483  ...          0.1135   
building 05      1.0000             1.0000  ...          0.5091   
building 06      0.0387             0.3131  ...          1.0000   

             Type05IsoOclu  Type06Area  Type06Integration  Type06Entropy  \
#                                                                          
building 01         0.0071      0.0562             0.6846         0.7077   
building 02         0.0017      0.1638             0.7126         0.3571   
building 03         0.1327      0.4380             0.9911         0.7883   
building 04         0.0924      0.4273             0.8558         0.7782   
building 05         0.9143      1.0000             1.0000         1.0000   
building 06         1.0000      0.0230             0.4560         0.5149   

             Type06Control  Type06Choice  Type06IsoArea  Type06IsoPerim  \
#                                                                         
building 01         0.5101        0.6811         0.0252          0.0578   
building 02         0.5217        0.2142         0.0040          0.0132   
building 03         0.9710        0.7387         0.0573          0.1119   
building 04         0.5072        0.8292         0.0356          0.0802   
building 05         1.0000        1.0000         1.0000          1.0000   
building 06         0.2319        0.2972         0.0040          0.0129   

             Type06IsoOclu  
#                           
building 01         0.0074  
building 02         0.0020  
building 03         0.0369  
building 04         0.0125  
building 05         1.0000  
building 06         0.0000  

[6 rows x 48 columns]

## Import Labels

Assign address for data table from labels to variable ('y' values).
This set represents the labels by which the quant data
will be trainned. In this case these are scores extracted
from a survey, using a phenomenological approach, about
the different buildings from 1 to 5.
The scores were separated into different fields such as
orientation, circulation, hierarchy of spaces, etc.

In [5]:
y_values = import_data('labels_example.xlsx', 'trainning_labels')

Print data table.

In [6]:
y_values

Hierarchy  Circulation  Efficiency  Proportion  Kitchen  Outdoor  \
#                                                                               
building 01       2.56         2.44        2.56        2.44     2.56     2.33   
building 02       3.56         3.78        3.78        3.89     3.33     2.89   
building 03       3.00         3.11        3.11        3.11     3.33     2.22   
building 04       2.33         2.67        3.00        3.22     3.22     2.00   
building 05       3.78         4.00        3.22        3.33     3.44     3.78   
building 06       3.56         3.44        3.44        3.67     3.89     4.11   

             Orientation  Overall  
#                                  
building 01         3.11     2.44  
building 02         3.56     3.56  
building 03         2.89     2.78  
building 04         2.89     2.56  
building 05         3.11     3.89  
building 06         3.67     3.56

## Import Testing Data

Read data table from excel file.

In [7]:
testing_data = import_data('quant_data_example.xlsx', 'testing_set')

Print data table.

In [8]:
testing_data

Type01Area  Type01Integration  Type01Entropy  Type01Control  \
#                                                                            
test building      0.5516             0.7514         0.9693         0.3516   

               Type01Choice  Type01IsoArea  Type01IsoPerim  Type01IsoOclu  \
#                                                                           
test building        0.4267         0.1355          0.1869         0.0066   

               Type02Area  Type02Integration  ...  Type05IsoPerim  \
#                                             ...                   
test building       0.763             0.6228  ...          0.5629   

               Type05IsoOclu  Type06Area  Type06Integration  Type06Entropy  \
#                                                                            
test building         0.0428      0.6003             0.7838         0.8933   

               Type06Control  Type06Choice  Type06IsoArea  Type06IsoPerim  \
#                                                                           
test building              1             1         0.0265          0.0635   

               Type06IsoOclu  
#                             
test building         0.0003  

[1 rows x 48 columns]

# ML Models (Elastic Net)

## Elastic Net Definition

In [9]:
def elastic_net(x, y, cross_validation):
    """This function returns a trainned model from a data set ('x') based on a label ('y').
    The variable 'round_integer' defines the precision for coefficients for each variable 'x' for every 'y'.
    The variable 'cross_validation' defines the number of group samples k for cross validation.
    """
    trainned_model_cross = ElasticNetCV(cv=cross_validation, random_state=0).fit(x, y)
    trainned_model = ElasticNet(alpha = trainned_model_cross.alpha_).fit(x, y)
    return trainned_model

## Prediction

Taking the label "Hierarchy" from the example labels, lets train the model and predict the score of the test building for "Hierarchy".

In [10]:
trainned_model = elastic_net(x_values, y_values.Hierarchy, 6)

Plot the coefficients of the trainned model

In [11]:
np.round(trainned_model.coef_, 4)

array([-0., -0., -0.,  0., -0., -0., -0.,  0.,  0., -0., -0., -0., -0.,
        0.,  0.,  0., -0., -0., -0.,  0., -0., -0., -0.,  0., -0., -0.,
       -0., -0., -0., -0., -0., -0.,  0.,  0., -0.,  0.,  0.,  0.,  0.,
        0.,  0., -0., -0.,  0., -0.,  0.,  0.,  0.])

Predict the score test building

In [12]:
trainned_model.predict(testing_data)

array([3.13166667])

# Conclusion

After trainning the model for the specific label "Hierarchy", the score for that label for the test building is 3.13 out of 5.00